File: content/snippets/scala3/interview_lcci_01_05_one_away.scala
object Lcci0105OneAway:
  def oneAway(a: String, b: String): Boolean =
    if (math.abs(a.length - b.length) > 1) {
      return false
    }

    val (s1, s2) = if (a.length <= b.length) (a, b) else (b, a)
    var i = 0
    var j = 0
    var found = false

    while (i < s1.length && j < s2.length) {
      if (s1(i) == s2(j)) {
        i += 1
        j += 1
      } else {
        if (found) {
          return false
        }
        found = true
        if (s1.length == s2.length) {
          i += 1
        }
        j += 1
      }
    }

    true

  @main def main(): Unit =
    println(s"pale vs ple -> ${oneAway("pale", "ple")}")
    println(s"pales vs pale -> ${oneAway("pales", "pale")}")
    println(s"pale vs bake -> ${oneAway("pale", "bake")}")

File: content/snippets/go/interview_lcci_01_05_one_away.go
package main

import "fmt"

func oneAway(a string, b string) bool {
	if len(a)-len(b) > 1 || len(b)-len(a) > 1 {
		return false
	}

	s1, s2 := a, b
	if len(s1) > len(s2) {
		s1, s2 = s2, s1
	}

	i, j := 0, 0
	found := false

	for i < len(s1) && j < len(s2) {
		if s1[i] == s2[j] {
			i++
			j++
			continue
		}

		if found {
			return false
		}
		found = true

		if len(s1) == len(s2) {
			i++
		}
		j++
	}

	return true
}

func main() {
	fmt.Printf("pale vs ple -> %t\n", oneAway("pale", "ple"))
	fmt.Printf("pales vs pale -> %t\n", oneAway("pales", "pale"))
	fmt.Printf("pale vs bake -> %t\n", oneAway("pale", "bake"))
}
